text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,486 @@
+the risks is that a condition may contain misspelled identifiers, or may use identifiers that are
+obsolete or have been removed from a product. For example:
+
+```rust
+#[cfg(feature = "widnows")]
+fn do_windows_thing() { /* ... */ }
+```
+
+The developer intended to test for the feature named `windows`. This could easily have been detected
+by `rustc` if it had known the set of all valid `feature` flags, not only the ones currently
+enabled.
+
+This RFC proposes adding new command-line options to `rustc`, which will allow Cargo (and other
+build tools) to inform `rustc` of the set of valid feature flags. Using conditions that are not
+valid will cause a diagnostic warning. This feature is opt-in, for backwards compatibility;
+if no valid configuration options are presented to `rustc` then no warnings are generated.
+
+# Motivation
+
+* Stronger assurance that large code bases are correct.
+* Protect against typos, bad merges, etc.
+* Detect dead code, typically caused by feature flags that have been removed from a crate's
+  manifest, but which still have `#[cfg(...)]` attributes that mention those features.
+
+# Guide-level explanation
+
+## Background
+
+Rust programs can use conditional compilation in order to modify programs based on the features a
+user has selected, the target CPU architecture, the target OS, or other parameters under control
+of the user. Rust programs may use conditional compilation in these ways:
+
+* By applying the `#[cfg(c)]` attribute to language elements, where `c` is a condition.
+* By applying the `#[cfg_attr(c, attr)]` attribute to language elements, where `c` is a conditional
+  and `attr` is an attribute to apply.
+* By using the `cfg!(c)` built-in macro, where `c` is a condition. The compiler replaces the macro
+  call with a `true` or `false` literal.
+
+A _condition_ can take one of two forms:
+
+* A single identifier, such as `#[cfg(test)]` or `#[cfg(linux)]`. These are Boolean conditions;
+  they are either enabled or disabled.
+* A condition may test whether a given value is present in a named list of values. For example,
+  `#[cfg(feature = "lighting")]` tests whether the `lighting` feature is enabled. Note that a given
+  condition name may have any number of enabled values; for example, it is legal to invoke
+  `rustc --cfg feature="lighting" --cfg feature="bump_maps"`.
+
+## Checking conditions names
+
+`rustc` can optionally verify that all condition names used in source code are valid. "Valid" is
+distinct from "enabled". A _valid_ condition is one that is allowed to appear in source code; the
+condition may be enabled or disabled, but it is still valid. An _enabled_ condition is one which has
+been specified with a `--cfg foo` or `--cfg 'foo="value"'` option.
+
+For example, `rustc` can detect this bug, where the `test` condition is misspelled as `tset`:
+
+```rust
+if cfg!(tset) {    // uh oh, should have been 'test'

```

Comment:
A future possibility would be to check for simple misspellings for the well-known condition names, even when `-cfg valid(...)` isn't used.

I guess this would catch most of the bugs this RFC wants to detect, but avoids the problem of false warnings caused by `-cfg`s from build scripts.
Implementation wise, maybe the logic from some diagnostics can be reused, like `help: a config with a similar name exists: 'test'`